rust/src/buffer.rs

```diff
@@ -20,18 +20,25 @@ use libc;
 use std::mem;
 use std::slice;
 
+use super::datatypes::*;
 use super::memory::*;
 
 /// Buffer<T> is essentially just a Vec<T> for fixed-width primitive types and the start of the
 /// memory region is aligned at a 64-byte boundary
-pub struct Buffer<T> {
+pub struct Buffer<T>
+where
+    T: ArrowPrimitiveType,
+{
     /// Contiguous memory region holding instances of primitive T
     data: *const T,
     /// Number of elements in the buffer
     len: i32,
```

Comment:
I'm curious why you don't make the length 64-bit?

Reply:
Naturally I would have used `usize` but I understood that Arrow requirs the type `i32` to be used for array lengths?

Reply:
The memory layout spec and the C++ implementation seem out of sync in that regard. The C++ implementation uses 64-bit ints for array sizes (it should perhaps use `size_t` or `ssize_t`). What *must* be 32-bit ints is offsets for list / string / binary arrays. @wesm